src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/maas/MaasProvider.java

```diff
@@ -0,0 +1,248 @@
+
+    @Autowired
+    private MaasProviderClientCache maasProviderClientCache;
+
+    @Override
+    public Set<Instance> createInstance(Infrastructure infrastructure, Instance instance) {
+
+        MaasClient maasClient = maasProviderClientCache.getMaasClient(infrastructure);
+
+        // Retrieve and convert the list of tags
+        List<org.ow2.proactive.connector.maas.data.Tag> tags = convertIaasTagsToMaasTags(tagManager.retrieveAllTags(instance.getOptions()));
+
+        // Initialize MAAS deployment polling
+        MaasClientPollingService maasPollingService = new MaasClientPollingService(maasClient,
+                                                                                   Integer.valueOf(instance.getNumber()));
+
+        // Start deployment(s) by ID or by resources
+        List<Future<Machine>> futureMachines;
+        if (instance.getId() != null) {
+            futureMachines = IntStream.rangeClosed(1, Integer.valueOf(instance.getNumber()))
+                                      .mapToObj(instanceIndexStartAt1 -> maasPollingService.deployMachine(instance.getId(),
+                                                                                                          instance.getInitScript()
+                                                                                                                  .getScripts()[0],
+                                                                                                          tags))
+                                      .collect(Collectors.toList());
+        } else {
+            futureMachines = IntStream.rangeClosed(1, Integer.valueOf(instance.getNumber()))
+                                      .mapToObj(instanceIndexStartAt1 -> maasPollingService.deployMachine(Integer.valueOf(instance.getHardware()
+                                                                                                                                  .getMinCores()),
+                                                                                                          Integer.valueOf(instance.getHardware()
+                                                                                                                                  .getMinRam()),
+                                                                                                          instance.getInitScript()
+                                                                                                                  .getScripts()[0],
+                                                                                                          tags))
+                                      .collect(Collectors.toList());
+        }
+
+        // Retrieve futures (blocking calls)
+        Set<Instance> instances = futureMachines.stream().map(futureMachine -> {
+            try {
+                return futureMachine.get();
+            } catch (InterruptedException | ExecutionException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)